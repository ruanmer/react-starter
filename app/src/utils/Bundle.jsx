import React, { PureComponent } from 'react';

export default class Bundle extends PureComponent {
  state = {
    Component: null
  }

  mounted = false

  componentDidMount () {
    this.mounted = true;

    this.load(this.props);
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.load !== this.props.load) {
      this.load(nextProps);
    }
  }

  componentWillUnmount () {
    this.mounted = false;
  }

  load (props) {
    this.setState({ Component: null });

    props.load(Component => {
      this.mounted && this.setState({ Component: Component.default });
    });
  }

  render () {
    const { Component } = this.state;

    if (Component) {
      const { load, ...props } = this.props;

      return <Component {...props} />;
    }

    return null;
  }
}
