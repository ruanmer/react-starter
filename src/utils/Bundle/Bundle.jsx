import React, { PureComponent } from 'react';

export default class Bundle extends PureComponent {
  state = {
    Component: null
  }

  componentDidMount () {
    this.mounted = true;

    this.load(this.props);
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.load !== this.props.load) {
      this.load(nextProps);
    }
  }

  componentWillUnmount () {
    this.mounted = false;
  }

  mounted = false

  load (props) {
    this.setState({ Component: null });

    props.load((Component) => {
      if (this.mounted) {
        this.setState({ Component: Component.default });
      }
    });
  }

  render () {
    const { Component } = this.state;

    if (Component) {
      const { load, ...props } = this.props;

      return <Component {...props} />;
    }

    return null;
  }
}
