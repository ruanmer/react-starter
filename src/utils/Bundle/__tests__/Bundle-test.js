/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow, mount } from 'enzyme';

import Bundle from '../Bundle';

describe('Bundle', () => {
  const MockComponent = () => <div>Mock Component</div>;

  it('should render null when no Component is returned', () => {
    const wrapper = shallow(
      <Bundle load={f => f} />
    );

    expect(wrapper.node).toMatchSnapshot();
  });

  it('should set mounted true when componentDidMount', () => {
    const wrapper = mount(
      <Bundle load={f => f} />
    );

    const wrapperInstance = wrapper.instance();

    expect(wrapperInstance.mounted).toBe(true);
  });

  it('should set mounted false when componentWillUnmount', () => {
    const wrapper = mount(
      <Bundle load={f => f} />
    );

    const wrapperInstance = wrapper.instance();

    wrapper.unmount();

    expect(wrapperInstance.mounted).toBe(false);
  });

  it('should call given load prop', () => {
    const loadComponent = jest.fn();

    mount(
      <Bundle load={loadComponent} />
    );

    expect(loadComponent).toHaveBeenCalledTimes(1);
  });

  it('should render the given Component', () => {
    const loadComponent = (callback) => {
      callback({ default: MockComponent });
    };

    const wrapper = mount(
      <Bundle load={loadComponent} />
    );

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call not load when componentWillReceiveProps and nextProps is equal to current props', () => {
    const spyLoad = jest.spyOn(Bundle.prototype, 'load');

    const loadComponent = (callback) => {
      callback({ default: MockComponent });
    };

    const wrapper = shallow(
      <Bundle load={loadComponent} />
    );

    wrapper.setProps({ load: loadComponent });

    expect(spyLoad).toHaveBeenCalledTimes(0);
  });

  it('should call load when componentWillReceiveProps and nextProps is not equal to current props', () => {
    const spyLoad = jest.spyOn(Bundle.prototype, 'load');

    const wrapper = shallow(
      <Bundle load={f => f} />
    );

    const loadComponent = (callback) => {
      callback({ default: MockComponent });
    };

    wrapper.setProps({ load: loadComponent });

    expect(spyLoad).toHaveBeenCalledTimes(1);
  });
});
