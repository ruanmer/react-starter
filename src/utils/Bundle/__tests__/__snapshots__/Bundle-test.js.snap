// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle should render null when no Component is returned 1`] = `null`;

exports[`Bundle should render the given Component 1`] = `"<div>Mock Component</div>"`;
